NuGetProtocolSamples/Program.cs

```diff
@@ -0,0 +1,133 @@
+            NuGetVersion packageVersion = new NuGetVersion("12.0.1");
+            using MemoryStream packageStream = new MemoryStream();
+
+            await resource.CopyNupkgToStreamAsync(
+                packageId,
+                packageVersion,
+                packageStream,
+                cache,
+                logger,
+                cancellationToken);
+
+            Console.WriteLine($"Downloaded package {packageId} {packageVersion}");
+
+            using PackageArchiveReader packageReader = new PackageArchiveReader(packageStream);
+            NuspecReader nuspecReader = await packageReader.GetNuspecReaderAsync(cancellationToken);
+
+            Console.WriteLine($"Tags: {nuspecReader.GetTags()}");
+            Console.WriteLine($"Description: {nuspecReader.GetDescription()}");
+        }
+
+        public static async Task GetPackageMetadataAsync()
+        {
+            ILogger logger = NullLogger.Instance;
+            CancellationToken cancellationToken = CancellationToken.None;
+
+            SourceCacheContext cache = new SourceCacheContext();
+            SourceRepository repository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
+            PackageMetadataResource resource = await repository.GetResourceAsync<PackageMetadataResource>();
+
+            IEnumerable<IPackageSearchMetadata> packages = await resource.GetMetadataAsync(
+                "Newtonsoft.Json",
+                includePrerelease: true,
+                includeUnlisted: false,
+                cache,
+                logger,
+                cancellationToken);
+
+            foreach (IPackageSearchMetadata package in packages)
+            {
+                Console.WriteLine($"Version: {package.Identity.Version}");
+                Console.WriteLine($"Listed: {package.IsListed}");
+                Console.WriteLine($"Tags: {package.Tags}");
+                Console.WriteLine($"Description: {package.Description}");
+            }
+        }
+
+        public static async Task SearchPackages()
+        {
+            ILogger logger = NullLogger.Instance;
+            CancellationToken cancellationToken = CancellationToken.None;
+
+            SourceRepository repository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
+            PackageSearchResource resource = await repository.GetResourceAsync<PackageSearchResource>();
+            SearchFilter searchFilter = new SearchFilter(includePrerelease: true);
+
+            IEnumerable<IPackageSearchMetadata> results = await resource.SearchAsync(
+                "json",
+                searchFilter,
+                skip: 0,
+                take: 20,
+                logger,
+                cancellationToken);
+
+            foreach (IPackageSearchMetadata result in results)
+            {
+                Console.WriteLine($"Found package {result.Identity.Id} {result.Identity.Version}");
+            }
+        }
```

Comment:
I love these examples...can we maybe include an example for an authenticated feed? 

Basically when doing the below
```cs
            SourceRepository repository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
```

Use the PackageSource type as an intermediary when getting the source repository because PackageSource has a password field.

Reply:
This is definitely a good idea! I also would like to add examples on how to upload and unlist packages. That being said, I think I'd want to do that in a separate change because:

1. All the examples in this change can be run with no configuration. It's not trivial to setup an authenticated feed
1. We're seeing lots of new tools around nuget.org ([FuGet](https://www.fuget.org/), [NuGetTrends](https://nugettrends.com/), [ToolGet](https://www.toolget.net/), but I don't know of any for private feeds. so I think focusing on public feeds first makes sense

Reply:
Makes sense...these 4 are sufficient then. :)